### DISCUSSION TITLE
Lowercasing package names

### DISCUSSION THREAD

**Discussion by isc-rsingh** [intersystems/ipm#326] [Ideas]
Current Community convention is to keep package names lowercase everywhere, including in the code ([reference](https://community.intersystems.com/post/naming-convention-objectscript-packages-classes-and-package-manager-modules-names)) 
- Agree that developers should use only lower-case when publishing to the community registry.
- Why must the names need to be coerced to lower case client side (once they are installed)?

- For HealthShare, packages have not been all lowercase and ISC cannot change the names of those modules as that would really complicate back-porting to adhocs.
  - For example, a module like HealthShare.ClientAPI.FederationRegistry, becomes healthshare.clientapi.federationregistry, which is a lot more difficult to read.
  - The names should match whatever their actual casing is if you have a registry (in our case Perforce) that doesn't coerce to lower case, which would maintain compatibility with existing behavior for the community since all packages in the community registry are lowercase. 
- Each registry should have the flexibility to specify their own naming convention

**Comment by daimor** [comment]
First of all, it's a common practice, for many package managers from many other languages. Nothing needs to be re-invented. It was chosen for reasons.

Case sensitivity is a way to typos. During installation, search, dependencies injection, and so on. It requires checking the uniquity of packages on many levels (pm and registry), and it requires much more work to make it stable.

If selecting between better readability of package names, and stability, I would choose last.

It is still possible, to keep readability, so, it would keep the original case during any display. But keep all the logic behind it in lowercase.

**Reply by isc-tleavitt** [reply]
I'd be a fan of compromising with Windows-style casing (display keeps original, matching is case-insensitive)

**Reply by isc-rsingh** [reply]
I'd be happy with that also.

**Reply by isc-kiyer** [reply]
Again I want to emphasise that I 100% agree that the registry should keep package names in all lower case. However, the zpm client, which installs the packages from the registry, need not coerce to lower case (since they should already be in lower case). 
For legacy registries (like we have in Perforce internally), we have mixed casing so want to preserve that.

I would also be fine with Tim's suggestion of storing in lower case on the client (if really needed. Seems unnecessary to me) and keeping original casing for display values.

**Comment by evshvarov** [comment]
What if we let the option to call them in any case?
E.g. zpm install Healhshare.Package and have it in the list only as
healthshare.package?
And of course with normalization to lowercase during publishing and install
operations under the hood?
Find/search will list all in lowercase for sure.

Ср, 15 июня 2022 г. в 19:49, isc-kiyer ***@***.***>:

**Reply by daimor** [reply]
At the moment, you can call as you wish when you try to install it, it will work, because of normalization to lower-case

**Reply by isc-kiyer** [reply]
The normalization to lower case only needs to happen on the registry server. Not the client side. Client side should always display in actual casing

**Reply by daimor** [reply]
focus only on Display

**Reply by isc-kiyer** [reply]
Even on the client side, what difference does it make when using just the community registry whether the names are stored in lower case or actual case?

**Reply by daimor** [reply]
logic inside has to be in lowercase
commands
```
install SomePackage
install somepackage
```
should always install the same package

**Reply by daimor** [reply]
Easier to keep it unique
so, there will no two different packages `SomePackage` and `somepackage`

**Reply by isc-kiyer** [reply]
You can have that behavior without coercing to lower case on the client side though. I understand the benefits of lower case but it is only relevant on the server side. Client side can keep whatever case the server responds with. That keeps compatibility with both community and HealthShare while coercing to lower case doesn't keep compatibility with HealthShare

**Reply by daimor** [reply]
So, SomePackage and somepackage, locally are different packages?

**Reply by isc-kiyer** [reply]
No the registry saves it as somepackage in the community case because the zpm registry converts to lower case. In the HealthShare case, yes they are different packages for our perforce-based repository.

**Reply by daimor** [reply]
It's only about that registry, but not all the packages have to be there
And does not matter where they are store, packages names have to be unique and case insensitive

**Comment by daimor** [comment]
Why, when the whole world already solved this issue, we keep speaking about it?
npm, lower-case only (since the beginning there was no such rule, but changed due to issues)
